### DISCUSSION TITLE
How can I use uuid in email verification url instead of id?

### DISCUSSION THREAD

**Discussion by Ifriqiya** [laravel/framework#44744] [Q&A]
I want to avoid exposing app metrics to users and this lead me to adding uuid to my user models and using it in routes. The email verification built into Laravel exposes user count by using user id in the verification url.

Is it possible for me to customise the authorize() method of the VerifyEmailRequest class to use uuid  instead of id?

**Comment by samaviam** [comment]
```
use Illuminate\Auth\Notifications\VerifyEmail;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\Config;
use Illuminate\Support\Facades\URL;

/**
* Bootstrap any application services.
*
* @return void
*/
public function boot()
{
   ...

    VerifyEmail::createUrlUsing(function ($notifiable) {
        return URL::temporarySignedRoute(
            'verification.verify',
            Carbon::now()->addMinutes(Config::get('auth.verification.expire', 60)),
            [
                'hash' => sha1($notifiable->getEmailForVerification()),
            ]
        );
    });
}
```

With this, you can customize the URL
Add this to the boot method of the AppServiceProvider class

**Reply by Ifriqiya** [reply]
Hi @Mahdi-Samavi, thanks for this. I tried it out but got a 403 error. My guess is that the framework VerifyEmail class requires an 'id' parameter which is missing in your suggestion. I still get a 403 when I add the 'id' variable and map it to  the user's 'uuid'. I think the Laravel Fortify VerifyEmailRequest class compares the route 'id' parameter to user()->getKey() and throws a 403 if it doesn't match. How do I customise this class?

**Reply by samaviam** [reply]
Hi @Ifriqiya
I think this [video](https://youtu.be/e-zhq5W0j1M) will solve your problem.

**Reply by Ifriqiya** [reply]
Thanks @Mahdi-Samavi, that worked.

**Comment by Ifriqiya** [comment]
I realised that all I needed to do to get email verification to use uuid was to override the getKey() method in my user model. But a side effect is that queries for user.id returns the uuid and my jobs fail with this message: 'No query results for model [App\Models\User].' because of this.

Hoe do I resolve this? I essentially need my queries to run with user.uuid instead of user.id.

**Reply by softmixt** [reply]
Go to your User model and add a new method , 

```php
public function getUuid(): string
{
     return $this->uuid;
}
```

then in your AuthServiceProvider.php add this to `boot`:

```php
VerifyEmail::createUrlUsing(function ($notifiable) {
            return URL::temporarySignedRoute(
                'verification.verify',
                Carbon::now()->addMinutes(Config::get('auth.verification.expire', 60)),
                [
                    'id' => $notifiable->getUuid(),
                    'hash' => sha1($notifiable->getEmailForVerification()),
                ]
            );
        });
```

Notice you have:  ` 'id' => $notifiable->getUuid(),` this way you will have in your email verification `uuid` instead `id` and you will keep the `id` to work as normal.
